prov/usnic/src/usdf_cq.c

```diff
@@ -428,6 +510,100 @@ usdf_cq_copy_soft_entry(void *dst, const struct usdf_cq_soft_entry *src,
 	return FI_SUCCESS;
 }
 
+static ssize_t
+usdf_cq_sread_common_soft(struct fid_cq *fcq, void *buf, size_t count, const void *cond,
+		int timeout, enum fi_cq_format format)
+{
+	struct usdf_cq *cq;
+	uint8_t *entry;
+	uint8_t *last;
+	struct usdf_cq_soft_entry *tail;
+	size_t entry_len;
+	size_t time_spent = 0;
+	ssize_t ret;
+
+	cq = cq_ftou(fcq);
+
+	switch (format) {
+	case FI_CQ_FORMAT_CONTEXT:
+		entry_len = sizeof(struct fi_cq_entry);
+		break;
+	case FI_CQ_FORMAT_MSG:
+		entry_len = sizeof(struct fi_cq_msg_entry);
+		break;
+	case FI_CQ_FORMAT_DATA:
+		entry_len = sizeof(struct fi_cq_data_entry);
+		break;
+	default:
+		USDF_WARN("unexpected CQ format, internal error\n");
+		return -FI_EOPNOTSUPP;
+	}
+
+	entry = buf;
+	last = entry + (entry_len * count);
+
+	while (1) {
+		/* progress... */
+		usdf_domain_progress(cq->cq_domain);
+
+		tail = cq->c.soft.cq_tail;
+
+		while (entry < last && tail != cq->c.soft.cq_head) {
+			if (tail->cse_prov_errno > 0) {
+				if (entry > (uint8_t *)buf)
+					break;
```

Comment:
This only breaks out of the inner `while` loop.  Don't you need to break out of the outer `while` loop instead (with `goto done_copying;` or similar)?